src/object.c

```diff
@@ -613,10 +613,16 @@ int isObjectRepresentableAsLongLong(robj *o, long long *llval) {
  * string. This happens because SDS strings tend to overallocate to avoid
  * wasting too much time in allocations when appending to the string. */
 void trimStringObjectIfNeeded(robj *o) {
-    if (o->encoding == OBJ_ENCODING_RAW &&
-        sdsavail(o->ptr) > sdslen(o->ptr)/10)
-    {
-        o->ptr = sdsRemoveFreeSpace(o->ptr, 0);
+    size_t len = sdslen(o->ptr);
+     /* A string may have free space in the following cases:
+     * 1. When an arg len is greater than PROTO_MBULK_BIG_ARG the query buffer may be used directly as the SDS string.
+     * 2. When utilizing the argument caching mechanism in Lua.
+     * 3. When the function is called from a module that has allocated a larger buffer than needed. */
+    if (len >= PROTO_MBULK_BIG_ARG || 
```

Comment:
i see in the previous PR you also made an attempt to use LUA_CMD_OBJCACHE_MAX_LEN and avoid unnecessary overheads in Lua scripts.
i admit it's a bit ugly, but i suppose it can have a big impact.
WDYT?

Reply:
I have added it back, it makes sense as the occurrence of free space in the Lua cache scenario is more of an exception than a regular event.